### PR TITLE
Fix editing of unsent messages by waiting for actual event id

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes vector-im/riot-web#11834
Supersedes https://github.com/matrix-org/matrix-react-sdk/pull/4220

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -287,7 +287,15 @@ export function MatrixClient(opts) {
                 _updatePendingEventStatus(room, eventToSend,
                                           EventStatus.SENDING);
             }
-            return _sendEventHttpRequest(self, eventToSend);
+            const promise = _sendEventHttpRequest(self, eventToSend);
+            if (room) {
+                // ensure we update pending event before the next scheduler run so that any listeners to event id
+                // updates on the synchronous event emitter get a chance to run first.
+                promise.then(res => {
```
Comment (foldleft) [inline src/client.js]: should this be returned, so anyone waiting on the promise also waits on this? if not, probably best to leave a comment that this is deliberate

File: src/client.js [hunk 2]
```diff
@@ -287,7 +287,16 @@ export function MatrixClient(opts) {
                 _updatePendingEventStatus(room, eventToSend,
                                           EventStatus.SENDING);
             }
-            return _sendEventHttpRequest(self, eventToSend);
+            let promise = _sendEventHttpRequest(self, eventToSend);
```
Comment (bwindels) [inline src/client.js]: nit: you could make this callback async, then you can just await here and don't need the promise reassignment below.

File: src/client.js [hunk 3]
```diff
@@ -281,22 +281,19 @@ export function MatrixClient(opts) {
...
-            let promise = _sendEventHttpRequest(self, eventToSend);
+            const res = await _sendEventHttpRequest(self, eventToSend);
             if (room) {
                 // ensure we update pending event before the next scheduler run so that any listeners to event id
                 // updates on the synchronous event emitter get a chance to run first.
-                promise = promise.then(res => {
-                    room.updatePendingEvent(eventToSend, EventStatus.SENT, res.event_id);
-                    return res;
-                });
+                room.updatePendingEvent(eventToSend, EventStatus.SENT, res.event_id);
```
Comment (bwindels) [inline src/client.js]: before you were returning the value of this call, is that not needed anymore?

**Comment by gdvine** [comment]
Nice work!

**Pull request merged**